### PR TITLE
feat: add filter, map, reduce options

### LINKED ISSUE #35: output file hash to manifest
What do you guys think about adding hash information to manifest file so we can make use of SRI?
https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity?
> Subresource Integrity (SRI) is a security feature that enables browsers to verify that files they fetch (for example, from a CDN) are delivered without unexpected manipulation. It works by allowing you to provide a cryptographic hash that a fetched file must match.

### LINKED ISSUE #46: Do not add code splitting bundles into manifest
I want to use `require.ensure` for separating some large modules as DLL.
But `webpack-manifest-plugin` add it into manifest 
so chunks are loading immediately

### LINKED ISSUE #55: Add reduce, initValue, map, filter options
- [x] filter (#66)
- [x] map
- [x] reduce
- [x] seed (#57)

-----

fix: #48 
```js
filter: ({chunk, file}) => {
  return chunk.isInitial();
}
```

fix: #25 

```js
initValue: {

} // will be used as the initValue for reduce, default to `{}`
```

fix #35 

```js
map: ({chunk, file}) => {
  return {
    file,
    hash: chunk.source.integrity
  };
}
```

```js
reduce: (manifest, {chunk, file}) => {
  return manifest;
}
```

-----
`map` and `reduce` can't be used at the same time

### PR DESCRIPTION
**Pull request by mastilver**
closes #46 
closes #55 
closes #35

### CODE REVIEW

File: spec/plugin.spec.js [hunk 1]
```diff
@@ -393,4 +393,29 @@ describe('ManifestPlugin', function() {
...
+        entry: {
+          nameless: './fixtures/nameless.js'
+        },
+        output: {
+          filename: '[name].[hash].js'
+        }
+      }, {
+        manifestOptions: {
+          filter: function(entry) {
+            return entry.isInitial;
```
Comment (mastilver) [inline spec/plugin.spec.js]: @a-x- Does that looks good to you?

File: spec/plugin.spec.js [hunk 2]
```diff
@@ -393,4 +393,85 @@ describe('ManifestPlugin', function() {
...
+        context: __dirname,
+        entry: './fixtures/file.js',
+        output: {
+          filename: '[name].js'
+        }
+      }, {
+        manifestOptions: {
+          reduce: function (manifest, file) {
+            manifest[file.name] = {
+              file:file.path,
```
Comment (a-x-) [inline spec/plugin.spec.js]: `:<space>`

File: README.md [hunk 3]
```diff
@@ -94,6 +95,68 @@ Default: `{}`
...
+
+Type: `function`
+
+Modify files details before the manifest is created.
+
+### `reduce`
+
+Type: `function`
+
+Create the manifest. It must return an `Object`.  The default one will take the name of the chunk as a key and the path of the file as the value.
```
Comment (a-x-) [inline README.md]: As I understand, we should note here: «custom reduce opt function replaces that default `reduce` logic:
```js
function (manifest, file) {
        manifest[file.name] = file.path;
        return manifest;
}
```

File: spec/plugin.spec.js (hunk 2 above)
Comment (a-x-) [inline spec/plugin.spec.js]: are we need eslint?

Comment (mastilver) [inline spec/plugin.spec.js]: We definitely need to setup eslint as some point

I personally have a preference for https://github.com/sindresorhus/xo , but really I don't mind ;)

File: README.md [hunk 4]
```diff
@@ -94,6 +95,68 @@ Default: `{}`
 
 A cache of key/value pairs to used to seed the manifest. This may include a set of [custom key/value](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json) pairs to include in your manifest, or may be used to combine manifests across compilations in [multi-compiler mode](https://github.com/webpack/webpack/tree/master/examples/multi-compiler). To combine manifests, pass a shared seed object to each compiler's ManifestPlugin instance.
 
+### `filter`
+
+Type: `function`
+
+Filter out files.
```
Comment (mastilver) [inline README.md]: I don't know, I don't really want to pollute the readme too much...

File: lib/plugin.js [hunk 5]
```diff
@@ -41,36 +44,73 @@ ManifestPlugin.prototype.apply = function(compiler) {
...
+
+
+    var manifest;
+    if (this.opts.reduce) {
+      manifest = files.reduce(this.opts.reduce, {});
+    } else {
+      manifest = files.reduce(function (manifest, file) {
+        manifest[file.name] = file.path;
+        return manifest;
+      }, {});
```
Comment (mastilver) [inline lib/plugin.js]: use `seed` instead of `{}`

File: spec/plugin.spec.js (hunk 2 above)
Comment (a-x-) [reply spec/plugin.spec.js]: is xo like [prettier](https://github.com/prettier/prettier)?

**Pull request merged**